### PR TITLE
 feat(text-field): character counter

### PR DESCRIPTION
**Pull request by TroyTae**
related to #817

### TODO
- [x] Create Component
- [x] Create Unit Test
- [x] Create Screenshot Test
- [x] Make Test Coverage 100%
- [x] Write README.md

### CODE REVIEW

**Comment by moog16** [comment]
@TroyTae is this ready for review? I checked out code and it failed to compile. Let me know when you're ready. Thanks! :)

**Comment by TroyTae** [comment]
@moog16 
Is there compile error?
Can you show me?
I have no error and unit test is passed.

And I need one help.
I don't understand about screenshot test well...
How can I fix screenshot test error?

**Comment by moog16** [comment]
@TroyTae looks like it is working now!

**Comment by moog16** [comment]
Can you add a screenshot tests to the main text-field and textarea screenshot tests? We need to see how it works with the helper text and with input coming into the text field.

**Comment by TroyTae** [comment]
@moog16 
Thank you for your advice!
But I seem I made a wrong code...
Maybe I need to study about puppeteer and docker.
I could understand and fix about unit test clearly.
But screenshot test isn't clear for me 😫

**Comment by moog16** [comment]
@TroyTae no worries - you need to install docker and go through a lot of steps. Where are you running into issues? Also I'm on Discord, so please reach out to me there if you need 1:1 help.

**Comment by TroyTae** [comment]
@moog16 
I'm ready for reviews!

**Comment by moog16** [comment]
Screen Shot 2019-05-23 at 10 30 25 AM

@TroyTae A few things:

- charCounter should be the same width of the text field. Can you change the width of the container so that it looks correct? 
- Adding the char counter doubles the amount of tests cases. I think to prevent this page from becoming too long, let's just create a single text field page similar to `text-field/refTest.tsx`. Then we don't have to change any thing. How does that sound?
- The second number shouldn't be 0. Can you change this to have a maxLength or something?

**Comment by TroyTae** [comment]
> charCounter should be the same width of the text field. Can you change the width of the container so that it looks correct?

I reflect it!
I set width inline style.
It looks good but I'll think about "Is this good code pattern?".

> Adding the char counter doubles the amount of tests cases. I think to prevent this page from becoming too long, let's just create a single text field page similar to text-field/refTest.tsx. Then we don't have to change any thing. How does that sound?

Yeah text-field screenshot has too long page.
But tests like refTest can not take every test case.
Is it ok?

> The second number shouldn't be 0. Can you change this to have a maxLength or something?

I reflect it!

> Please add a screenshot test case with helper text as well. We want to ensure that the helper text and charCounter can exist together.

It looks there are some case that shows helper-text and character-counter together.
image

**Comment by moog16** [comment]
>But tests like refTest cannot take every test case.
>Is it ok?
Yes I think that's fine. We only need to see a few cases working for the character counter with the text field.

Looks like its working well. Just gotta get the screenshot tests in order :)

**Comment by TroyTae** [comment]
@moog16 
Oh I see
I added characterCounter.tsx for screenshot test!

File: packages/chips/Chip.tsx [hunk 1]
```diff
@@ -163,13 +163,7 @@ export class Chip extends React.Component<ChipProps, ChipState> {
...
       getCheckmarkBoundingClientRect: () => {
         const {chipCheckmark} = this.props;
-        if (
-          !(
-            chipCheckmark &&
-            chipCheckmark.props &&
-            chipCheckmark.props.getBoundingClientRect
-          )
-        ) {
+        if (!(chipCheckmark && chipCheckmark.props && chipCheckmark.props.getBoundingClientRect)) {
```
Comment (moog16) [inline packages/chips/Chip.tsx]: can you revert this change please.

File: packages/text-field/character-counter/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# React Text Field Character Counter
+
+MDC React Text Field Character Counter is a React Component which uses [MDC Text Field Character Counter](https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield/character-counter)'s CSS and foundation JavaScript.
```
Comment (moog16) [inline packages/text-field/character-counter/README.md]: "...CSS and foundation JavaScript." --> "...Sass and Foundational JavaScript logic."

File: packages/text-field/character-counter/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  );
+}
+```
+
+## Props
+
+Prop Name | Type | Description
+--- | --- | ---
+className | String | CSS classes for element.
+template | String | You can set custom template. See below
```
Comment (moog16) [inline packages/text-field/character-counter/README.md]: "See below" --> `[See below](#custom-template)`

File: packages/text-field/character-counter/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+``` js
+import React from 'react';
+import TextField, {CharacterCounter, Input} from '@material/react-text-field';
+
+class MyApp extends React.Component {
+  render() {
+    return (
+      <TextField characterCounter={<CharacterCounter template='${count} : ${maxLength}' />}>
+        <Input value={'Happy Coding!'} maxLength={140} />
```
Comment (moog16) [inline packages/text-field/character-counter/README.md]: change to `<Input value={this.state.value} maxLength={140} onChange={(e) => this.setState({value: e.target.value})} />`.

File: packages/text-field/character-counter/README.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+For example, default template is `${count} / ${maxLength}`, so it appears `0 / 140`.  
+If you set template as `${count} : ${maxLength}`, it appears `0 : 140`.  
+
+### Sample
+
+``` js
+import React from 'react';
+import TextField, {CharacterCounter, Input} from '@material/react-text-field';
+
+class MyApp extends React.Component {
```
Comment (moog16) [inline packages/text-field/character-counter/README.md]: add

```
state = {value: 'Happy Coding!'};
```

File: packages/text-field/character-counter/README.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+## Props
+
+Prop Name | Type | Description
+--- | --- | ---
+className | String | CSS classes for element.
+template | String | You can set custom template. See below
+
+## Custom Template
+
+CharacterCounter provides customizing by setting the `template` prop in CharacterCounter.
```
Comment (moog16) [inline packages/text-field/character-counter/README.md]: Change to
```
CharacterCounter provides customization with the `template` prop in CharacterCounter. The `template` prop accepts the `${count}` and `${maxLength}` arguments. The default template is `${count} / ${maxLength}`, so it appears `0 / 140`.  If you set template as `${count} : ${maxLength}`, it appears as `0 : 140`.  
```

File: packages/text-field/index.tsx [hunk 7]
```diff
@@ -278,10 +281,11 @@ class TextField<
     };
   }
 
-  inputProps(child: React.ReactElement<InputProps<T>>) {
+  get inputProps() {
     // ref does exist on React.ReactElement<InputProps<T>>
     // @ts-ignore
-    const {props} = child;
+    const {props} = React.Children.only(this.props.children);
```
Comment (moog16) [inline packages/text-field/index.tsx]: Why did you add this .only here? I don't think we should assume that it can only be one child. This will be a breaking change.

File: packages/text-field/index.tsx [hunk 8]
```diff
@@ -394,12 +407,13 @@ class TextField<
 
   renderHelperLine(
     helperText?: React.ReactElement<HelperTextProps>,
-    characterCounter?: React.ReactElement<any>
+    characterCounter?: React.ReactElement<CharacterCounterProps>
   ) {
+    const el = this.textFieldElement.current;
     return (
-      <div className={cssClasses.HELPER_LINE}>
+      <div className={cssClasses.HELPER_LINE} style={{width: el ? `${el.clientWidth}px` : 'auto'}}>
```
Comment (moog16) [inline packages/text-field/index.tsx]: We shouldn't have `style` here - this should use the MDC Web's styles. Is this for the screenshot tests?

File: test/screenshot/text-field/TextFieldPermutations.tsx [hunk 9]
```diff
@@ -25,37 +27,39 @@ const textFields = (variantProps: {variant?: string}) => {
...
-                  value
-                );
-                const key = `${iconKey}-${denseKey}-${disabledKey}-${helperTextKey}-${isRequiredKey}--${rtlKey}`;
-                const hasIcon =
-                  iconKey === 'leadingIcon' || iconKey === 'trailingIcon';
-                if (hasIcon && variantProps.variant === 'fullWidth') {
-                  return;
-                }
-                return <TestField {...props} key={key} id={key} />;
+                return characterCounterMap.map((characterCounter: CharacterCounterMapType | {}) => {
```
Comment (moog16) [inline test/screenshot/text-field/TextFieldPermutations.tsx]: There are a ton of warnings on this page. Let's just revert this file so we keep the screenshot test page smaller.

File: test/screenshot/text-field/characterCounter.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import TextField, {CharacterCounter, Input} from '../../../packages/text-field';
+
+class CharacterCounterTextField extends React.Component<{}, {value: string}> {
```
Comment (moog16) [inline test/screenshot/text-field/characterCounter.tsx]: Please add the textarea variant

File: test/screenshot/text-field/attributesMap.tsx [hunk 11]
```diff
@@ -23,6 +24,11 @@ const helperTextMap = [
   {helperText: <HelperText persistent>Help me</HelperText>},
   {helperText: <HelperText validation>Error message</HelperText>},
 ];
+const characterCounterMap = [
+  {},
+  {characterCounter: <CharacterCounter />},
+  {characterCounter: <CharacterCounter template='현재(${count}) / 최대(${maxLength})' />},
```
Comment (moog16) [inline test/screenshot/text-field/attributesMap.tsx]: Add this case to the `test/screenshot/text-field/characterCounter.tsx` file.

File: test/screenshot/text-field/textArea.tsx [hunk 12]
```diff
@@ -19,25 +21,28 @@ const textareaFields = textareaVariants.map((variant) => {
...
-              dense,
-              disabled,
-              helperText,
-              isRtl
-            );
-            const key = `textarea-${JSON.stringify(props)}${helperTextKey}`;
-            return (
-              <TestField variant='textarea' {...props} key={key} id={key} />
-            );
+            return characterCounterMap.map((characterCounter: CharacterCounterMapType | {}) => {
```
Comment (moog16) [inline test/screenshot/text-field/textArea.tsx]: revert this too, and just add a case to the `test/screenshot/text-field/characterCounter.tsx` file.

File: test/unit/text-field/character-counter/index.test.tsx [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+      <Input />
+    </TextField>
+  );
+  assert.equal(
+    wrapper.childAt(0).childAt(0).getDOMNode(),
+    wrapper.find('.mdc-text-field-character-counter').getDOMNode()
+  );
+});
+
+test('Useless test for test coverage', () => {
```
Comment (moog16) [inline test/unit/text-field/character-counter/index.test.tsx]: Add a comment just saying that MDC React doesn't need to implement this

File: packages/text-field/index.tsx (hunk 7 above)
Comment (TroyTae) [reply packages/text-field/index.tsx]:
```
renderInput() {
  const child: React.ReactElement<InputProps<T>> = React.Children.only(this.props.children);
  const props = this.inputProps(child);
  return React.cloneElement(child, props);
}
```
This is original code of renderInput function.
So I thought text-field was receiving only one intentionally.

File: packages/text-field/index.tsx (hunk 8 above)
Comment (TroyTae) [reply packages/text-field/index.tsx]: I thought there is a spec that helper-line width.
But it looks my misunderstanding...
Thank you!

File: packages/text-field/index.tsx (hunk 7 above)
Comment (moog16) [reply packages/text-field/index.tsx]: ah ok - if it was there before, then that is fine :)

File: test/screenshot/golden.json [hunk 14]
```diff
@@ -22,13 +22,14 @@
...
   "tab-scroller": "468866dd0c222b36b55485ab44a5760133a4ddfb2a6cf81e6ae4672d7e02a447",
-  "text-field/helper-text": "59604d0f39e0846fc97aae7573d317dded215282a677e4641c5e33426e3a2a1e",
-  "text-field/icon": "0bbc8c762d27071e55983e5742548d166864b6fcebc0b9f1e413523fb93b7075",
-  "text-field/textArea": "dde78e3f154a8b910a989f8ce96e320e7ad2b3e199e6e7a81034174c598cbd9d",
-  "text-field/standard": "43c25e9eaadb2cd0f7ceeac939d3416344504d92baaca5ed8be02df84c2bba79",
-  "text-field/fullWidth": "6555b4398509aa79f2e6ae1ef3be678a9a877984c4aaf77804afc01fe4fc50a6",
-  "text-field/outlined": "a6f22c45fe20e8dab39c39711cbec6c249933b7c655d4dff39d4a635c846acc6",
-  "text-field/refTest": "0310e81ea870fb0e6e5968d0e1567e22b0e952aeeff9653d3dfe8d9b5b1f5588",
+  "text-field/character-counter": "3b9f1d0e66b0a717b1b8479c99efdde6ac9beddc536f2f3f53a814ad9d430bd6",
+  "text-field/helper-text": "7d9307a22892706762c976f4aa2973e9499270b382dbef9194e7b4a6e698adaf",
```
Comment (moog16) [inline test/screenshot/golden.json]: revert all of these...I will have the charcounter golden soon

Comment (moog16) [reply test/screenshot/golden.json]:
```
"text-field/characterCounter": "5c7afb850e7b34b65b64bb81fc2106d1b04d7af0fc14deca26938b479881373"
"text-field/character-counter": "fccecc3a8d1b392571abf1c28615cf2a8f4a29b3e6f50f6030ccf0464e996bd9"
```

**Comment by moog16** [comment]
Just need screenshots to pass...looks good!

**Comment by moog16** [comment]
@TroyTae please sign CLA

**Comment by moog16** [comment]
running tests for https://github.com/material-components/material-components-web-react/pull/897

**Comment by TroyTae** [comment]
I signed it!

**Comment by TroyTae** [comment]
I signed it!

**Comment by moog16** [comment]
@TroyTae looks like unit tests are failing. Some typescript error

**Comment by TroyTae** [comment]
I think there was an error in the merge process.
Now I fix it!

**Comment by moog16** [comment]
@TroyTae I updated branch to rc0.14.0. Please fix conflicts.

**Comment by TroyTae** [comment]
@moog16 
I fixed it!

**Comment by moog16** [comment]
@TroyTae can you try creating a new branch from your rc0.13.0 branch and opening a PR? I'm not sure why your branch isn't picking up the changes in the screenshot tests. Possibly some kind of name collision could be happening? I can confirm there is a text-field/characterCounter image https://storage.cloud.google.com/screenshot-uploads/text-field/characterCounter/5c7afb850e7b34b65b64bb81fc2106d1b04d7af0fc14deca26938b4798813738.golden.png?organizationId=433637338589&_ga=2.214246231.-1322722271.1547788303

**Comment by TroyTae** [comment]
@moog16 
First I combined two screenshot test file.
If this error's cause is name conflict, I guess this change can solve them.

File: test/screenshot/text-field/variants.tsx [hunk 15]
```diff
@@ -5,5 +5,4 @@ export default [
   'textArea',
   'refTest',
   'autoFocus',
-  'characterCounter'
```
Comment (moog16) [inline test/screenshot/text-field/variants.tsx]: I actually think we wanted this but let's just move this into another PR after we get this one in.

Comment (TroyTae) [reply test/screenshot/text-field/variants.tsx]: Ah then we don't need to create another PR, right?

Comment (moog16) [reply test/screenshot/text-field/variants.tsx]: Correct :)

**Comment by moog16** [comment]
signed in https://github.com/material-components/material-components-web-react/pull/861#issuecomment-497165346

**Comment by moog16** [comment]
@TroyTae please run `npm run fix` and fix all prettier/lint warnings. Then we're good to go!

**Comment by TroyTae** [comment]
@moog16 
I don't know about `npm run fix`.
So I found each one and fixed it 😂 
Anyway I fixed it!

**Comment by moog16** [comment]
Thanks!
